File: 02_flashlight/python/flashlight2.py
import RPi.GPIO as GPIO

ledPin = 11
buttonPin = 12
ledState = False


def setup():
    GPIO.setmode(GPIO.BOARD)

    GPIO.setup(ledPin, GPIO.OUT)
    GPIO.output(ledPin, GPIO.LOW)
    print('Pin #%d set to OUTPUT mode.' % ledPin)

    GPIO.setup(buttonPin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
    print('Pin #%d set to INPUT mode (pull-up).' % buttonPin)


def onBtnPress(channel):
    global ledState

    print('You pressed the button! (GPIO %d)' % channel)
    ledState = not ledState

    if ledState:
        print('Turning LED on...')
    else:
        print('Turning LED on...')

    GPIO.output(ledPin, ledState)


def loop():
    GPIO.add_event_detect(buttonPin, GPIO.FALLING,
                          callback=onBtnPress, bouncetime=300)

    while True:
        pass


def destroy():
    GPIO.output(ledPin, GPIO.LOW)
    GPIO.cleanup()


if __name__ == '__main__':
    setup()

    try:
        loop()
    except KeyboardInterrupt:
        destroy()

File: 02_push-button-led/c/flashlight2.c
#include <wiringPi.h>
#include <stdio.h>

#define ledPin 0
#define buttonPin 1

int ledState = LOW;
int buttonState = HIGH;
int lastBtnState = HIGH;
int tempBtnState;
long lastChangeTime;
long debounceThreshold = 50;

int main(void)
{
	/* Required to initialize WiringPi */
	wiringPiSetup();

	pinMode(ledPin, OUTPUT);
	printf("Pin #%d set to OUTPUT mode.\n", ledPin);

	pinMode(buttonPin, INPUT);
	pullUpDnControl(buttonPin, PUD_UP);
	printf("Pin #%d set to INPUT mode (pull-up).\n", buttonPin);

	while (TRUE)
	{
		tempBtnState = digitalRead(buttonPin);

		if (tempBtnState != lastBtnState)
		{
			lastChangeTime = millis();
		}

		if (millis() - lastChangeTime >= debounceThreshold)
		{
			if (tempBtnState != buttonState)
			{
				buttonState = tempBtnState;

				if (buttonState == LOW)
				{
					printf("You pressed the button!\n");
					ledState = !ledState;

					if (ledState)
					{
						printf("Turning LED on...\n");
					}
					else
					{
						printf("Turning LED off...\n");
					}
				}
			}
		}

		digitalWrite(ledPin, ledState);
		lastBtnState = tempBtnState;
	}

	return 0;
}
